pydicom/dataset.py

```diff
@@ -1197,6 +1249,36 @@ def __setitem__(self, key, value):
                 data_element.private_creator = self[private_creator_tag].value
         dict.__setitem__(self, tag, data_element)
 
+    def _slice_dataset(self, start, stop, step):
+        """Yield the element tags in the Dataset that match the slice.
+
+        Parameters
+        ----------
+        start : int or None
+            The slice's starting element tag value.
+        stop : int or None
+            The slice's stopping element tag value.
+        step : int or None
+            The slice's step size.
+
+        Yields
+        ------
+        pydicom.tag.Tag
+            The tags in the Dataset that meet the conditions of the slice.
+        """
+        all_tags = sorted(self.keys())
+        if start is None:
+            start = all_tags[0]
+        if stop is None:
+            stop = all_tags[-1] + 1
+
+        for tag in all_tags:
+            if Tag(start) <= tag < Tag(stop):
+                if step is not None and tag % step == 0:
+                    yield tag
+                elif step is None:
+                    yield tag
```

Comment:
Also, since the keys are sorted, the loop does not have to go through all tags.  At the very least it could break once `tag` was greater than `stop`.